### PR TITLE
Modify Irmin to work with the refactoring of Repr's encoding

### PR DESCRIPTION
**Pull request by mattiasdrp**
[Repr's PR](https://github.com/mirage/repr/pull/57)

There's a lot of dirty code and sub-optimal implementations but I need insights to keep going

### CODE REVIEW

File: src/irmin/utils.ml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+let size_of t = Repr.unstage (Repr.size_of t)
+let to_bin_string t = Repr.(unstage (to_bin_string t))
+let pre_hash t = Repr.(unstage (pre_hash t))
+
+let old_pre_hash t =
```
Comment (Ngoguey42) [inline src/irmin/utils.ml:5]: Since `old_pre_hash` is used only once, maybe you could move it back to `hash.ml`

**Review by Ngoguey42 — COMMENTED** [review]
My naive comments :)

File: src/irmin-pack/IO.ml [hunk 2]
```diff
@@ -70,6 +70,11 @@ module Unix : S = struct
     t.offset <- t.offset ++ len;
     if t.offset -- t.flushed > auto_flush_limit then flush t
 
+  let append_directly t size encoding =
+    let byt = Bytes.create size in
```
[the remark this answers is not in the document]
Comment (mattiasdrp) [reply src/irmin-pack/IO.ml:74]: This is actually the main focus point for me. What do we do with `IO.t` now that we can write directly in its buffer? @samoht @CraigFe

File: src/irmin-pack/IO.ml [hunk 3]
```diff
@@ -70,6 +70,11 @@ module Unix : S = struct
...
+    let off = encoding byt 0 in
+    append t (Bytes.sub_string byt 0 off)
```
[the remark this answers is not in the document]
Comment (mattiasdrp) [reply src/irmin-pack/IO.ml:76]: Let's hope it's just for now :-)

File: src/irmin-pack/content_addressable.ml [hunk 4]
```diff
@@ -237,6 +237,13 @@ struct
         Lwt.return r)
 
     let auto_flush = 1024
+    let size_of = Irmin.Type.unstage (Irmin.Type.size_of Val.t)
+    let to_bin_string = Irmin.Type.(unstage (to_bin_string Val.t))
+
+    let size_of_v t =
+      match size_of t with
+      | Some n -> n
+      | None -> String.length (to_bin_string t)
```
[the remark this answers is not in the document]
Comment (mattiasdrp) [reply src/irmin-pack/content_addressable.ml:246]: It shouldn't happen a lot but that's not the main problem with computing the size of our value from the limited benchmarks I did

File: src/irmin/hash.ml [hunk 5]
```diff
@@ -63,8 +63,7 @@ module Typed (K : S) (V : Type.S) = struct
 
   type value = V.t
 
-  let pre_hash = Type.unstage (Type.pre_hash V.t)
-  let hash v = K.hash (pre_hash v)
+  let hash v = K.hash (Utils.old_pre_hash V.t v)
```
[the remark this answers is not in the document]
Comment (mattiasdrp) [reply src/irmin/hash.ml:66]: Yeah, I need to clean these partial calls

**Pull request closed without merging**